catalog/core/catalog-core-injectattribute/src/main/java/ddf/catalog/data/inject/AttributeInjectorImpl.java

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.data.inject;
+
+import static org.apache.commons.lang.Validate.notNull;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.stream.Collectors;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.AttributeInjector;
+import ddf.catalog.data.AttributeRegistry;
+import ddf.catalog.data.InjectableAttribute;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.MetacardType;
+import ddf.catalog.data.impl.MetacardImpl;
+import ddf.catalog.data.impl.MetacardTypeImpl;
+
+public class AttributeInjectorImpl implements AttributeInjector {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AttributeInjectorImpl.class);
+
+    private final AttributeRegistry attributeRegistry;
+
+    private final Set<String> globalInjections = new CopyOnWriteArraySet<>();
+
+    private final Map<String, Set<String>> metacardTypeInjections = new ConcurrentHashMap<>();
+
+    public AttributeInjectorImpl(AttributeRegistry attributeRegistry) {
+        this.attributeRegistry = attributeRegistry;
+    }
+
+    private BundleContext getBundleContext() {
+        return FrameworkUtil.getBundle(getClass())
+                .getBundleContext();
+    }
+
+    public void bind(ServiceReference<InjectableAttribute> reference) {
```

Comment:
Since you never use the `ServiceReference` for anything other than looking up the `ServiceObject` you could actually listen for the `ServiceObjects` instead. I'll point it out in the blueprint.